### ISSUE TITLE
Captcha handling support

### ISSUE THREAD

**Issue by marzavec** [hack-chat/hackchat-client#23]
So, the current behavior for [captcha handling](https://github.com/hack-chat/hackchat-client/blob/master/app/containers/CommunicationProvider/reducer.js#L126) is to yeet the event into redux, shrug and call it a day.

This essentially means that captchas are not supported.

Captcha event are going to become part of normal spec, both in protecting channels from bots as well as claiming channels. Support for this event will need to be designed with both of these situations in mind. A model will need to be created to handle the display and input, some kind of callback solution added to "finish" the captcha.

Display compatibility (mostly on mobile) may benefit from using a canvas to render the captcha challenge text. . .

**Comment by KnowsCount** [comment]
well.. the only problem with CAPCHA is the fact that it's not friendly to network environments in China...

**Comment by marzavec** [comment]
Do you mean that the use Roman characters may cause issues with users who do not natively use Roman characters?

**Comment by KnowsCount** [comment]
Ah, ignore me. I thought you were talking about google's reCAPTCHA. Cos China banned Google

**Comment by marzavec** [comment]
Oh, for sure. Hack.chat strives to be entirely self-hosted, which means no external calls. The current captcha system uses [ascii-captcha](https://www.npmjs.com/package/ascii-captcha) to generate a captcha.

**Comment by xjzh123** [comment]
I wanna doubt the client displaying captcha. I dont know events and callbacks, but displaying has an impact on me.
I find captcha very hard even on my pc, and impossible on my phone.
It is just in a <pre> element, isn't it? I dont want to be unfriendly, but I am sure that the only solution known by me (and most of my friends) to captcha was to paste it in a notepad. Very few people say they are landscape mobile users and can pass captcha. I don't know any PC user who can pass captcha.
I guess I have made a display solution, though the way to do it may be rough. I don't use canvas, but SVG instead.
```
    captcha: function (args) {
		const NS = 'http://www.w3.org/2000/svg'

		let messageEl = document.createElement('div');
		messageEl.classList.add('message', 'info');


		let nickSpanEl = document.createElement('span');
		nickSpanEl.classList.add('nick');
		messageEl.appendChild(nickSpanEl);

		let nickLinkEl = document.createElement('a');
		nickLinkEl.textContent = '#';
		nickSpanEl.appendChild(nickLinkEl);

		let pEl = document.createElement('p')
		pEl.classList.add('text')

		let lines = args.text.split(/\n/g)

		// Core principle: In SVG text can be smaller than 12px even in Chrome.
		let svgEl = document.createElementNS(NS, 'svg')
		svgEl.setAttribute('white-space', 'pre')
		svgEl.style.backgroundColor = '#4e4e4e'
		svgEl.style.width = '100%'

		// In order to make 40em work right.
		svgEl.style.fontSize = `${$('#messages').clientWidth / lines[0].length * 1.5}px`
		// Captcha text is about 41 lines.
		svgEl.style.height = '41em'

		// I have tried `white-space: pre` but it didn't work, so I write each line in individual text tags.
		for (let i = 0; i < lines.length; i++) {
			let line = lines[i]
			let textEl = document.createElementNS(NS, 'text')
			textEl.innerHTML = line

			// In order to make it in the right position. 
			textEl.setAttribute('y', `${i + 1}em`)

			// Captcha text shouldn't overflow #messages element, so I divide the width of the messages container with the overvalued length of each line in order to get an undervalued max width of each character, and than multiply it by 2 (The overvalued aspect ratio of a character) because the font-size attribute means the height of a character. 
			textEl.setAttribute('font-size', `${$('#messages').clientWidth / lines[0].length * 1.5}px`)
			textEl.setAttribute('fill', 'white')

			// Preserve spaces.
			textEl.style.whiteSpace = 'pre'

			svgEl.appendChild(textEl)
		}

		pEl.appendChild(svgEl)

		messageEl.appendChild(pEl);
		$('#messages').appendChild(messageEl);

		window.scrollTo(0, document.body.scrollHeight);
	}
```
It works very well on me, though I don't know why you can normally read captcha on PC, so this may be improper on your devices.
btw, I think `O` and `0` are a bit ambiguous in captcha.

**Comment by marzavec** [comment]
Fixed with 37a9df1a97ac2384ed62c67ea5c6db2dfea631c4
